File: go/engine.go
package maqa

import (
	"math/rand"
	"sort"
)

// MAQAEngine filters candidate brokers and returns the final descending ranking.
type MAQAEngine struct {
	Config     Config
	Calculator *ScoreCalculator
}

// NewEngine creates a ranking engine.
func NewEngine(config Config) (*MAQAEngine, error) {
	calculator, err := NewScoreCalculator(config)
	if err != nil {
		return nil, err
	}
	return &MAQAEngine{Config: config, Calculator: calculator}, nil
}

// NewDefaultEngine creates a ranking engine with default parameters.
func NewDefaultEngine() (*MAQAEngine, error) {
	return NewEngine(DefaultConfig())
}

// IsEligible checks whether a broker is allowed to participate in ranking.
func (e MAQAEngine) IsEligible(broker Broker) bool {
	return broker.IsEligible && broker.SLAOK
}

// Rank is the main entry point of the Go MAQA implementation: filter, score, and sort by noisy_score descending.
func (e MAQAEngine) Rank(brokers []Broker, lead Lead, context RankingContext, rng *rand.Rand) RankingResult {
	ranked := make([]RankedBroker, 0, len(brokers))
	for _, broker := range brokers {
		if !e.IsEligible(broker) {
			continue
		}
		ranked = append(ranked, RankedBroker{
			Broker: broker,
			Score:  e.Calculator.CalcScoreBreakdown(broker, lead, context, rng),
		})
	}
	sort.Slice(ranked, func(i int, j int) bool {
		return ranked[i].Score.NoisyScore > ranked[j].Score.NoisyScore
	})
	return RankingResult{RankedBrokers: ranked}
}

File: python/maqa/engine.py
from __future__ import annotations

from random import Random

from .config import MAQAConfig
from .formulas import ScoreCalculator
from .models import Broker, Lead, RankedBroker, RankingContext, RankingResult


class MAQAEngine:
    # Filter candidates and produce the final ranking, with the first item as the default choice.
    def __init__(self, config: MAQAConfig | None = None, calculator: ScoreCalculator | None = None) -> None:
        self.config = config or MAQAConfig()
        self.calculator = calculator or ScoreCalculator(self.config)

    @staticmethod
    def is_eligible(broker: Broker) -> bool:
        return broker.is_eligible and broker.sla_ok

    def rank(
        self,
        brokers: tuple[Broker, ...],
        lead: Lead,
        context: RankingContext,
        rng: Random | None = None,
    ) -> RankingResult:
        # Main entry point: filter, score, and return brokers in descending score order.
        ranked_brokers = tuple(
            RankedBroker(
                broker=broker,
                score=self.calculator.calc_score_breakdown(broker, lead, context, rng),
            )
            for broker in brokers
            if self.is_eligible(broker)
        )
        return RankingResult(
            ranked_brokers=tuple(
                sorted(ranked_brokers, key=lambda item: item.score.noisy_score, reverse=True)
            )
        )
